opm/simulators/flow/FlowProblemBlackoil.hpp

```diff
@@ -1354,11 +1354,15 @@ class FlowProblemBlackoil : public FlowProblem<TypeTag>
                     dofFluidState.setSaturation(FluidSystem::gasPhaseIdx,
                                                 gasSaturationData[dofIdx]);
             }
-            if (FluidSystem::phaseIsActive(FluidSystem::oilPhaseIdx))
-                dofFluidState.setSaturation(FluidSystem::oilPhaseIdx,
-                                            1.0
-                                            - waterSaturationData[dofIdx]
-                                            - gasSaturationData[dofIdx]);
+            if (FluidSystem::phaseIsActive(FluidSystem::oilPhaseIdx)) {
+                const Scalar soil = 1.0 - waterSaturationData[dofIdx] - gasSaturationData[dofIdx];
+                if (soil < smallSaturationTolerance_) {
+                    dofFluidState.setSaturation(FluidSystem::oilPhaseIdx, 0.0);
```

Comment:
Instead for adding a tolerance check for soil here you could consider checking if rv < rvSat? In the end I dont think it should lead to a big differences from what you suggest.